runtime/signals.c

```diff
@@ -277,40 +278,35 @@ void caml_update_young_limit (void)
 void caml_request_major_slice (void)
 {
   Caml_state->requested_major_slice = 1;
-  caml_set_something_to_do();
+  caml_set_action_pending();
 }
 
 void caml_request_minor_gc (void)
 {
   Caml_state->requested_minor_gc = 1;
-  caml_set_something_to_do();
+  caml_set_action_pending();
 }
 
-value caml_check_gc_without_async_callbacks(value extra_root)
+void caml_process_pending_callbacks(void)
 {
-  CAMLparam1 (extra_root);
-  if (Caml_state->requested_major_slice || Caml_state->requested_minor_gc){
-    CAML_INSTR_INT ("force_minor/check_gc_without_async_callbacks@", 1);
-    caml_gc_dispatch();
-  }
-  CAMLreturn (extra_root);
-}
-
-CAMLexport value caml_check_urgent_gc (value extra_root)
-{
-  CAMLparam1 (extra_root);
   caml_something_to_do = 0;
```

Comment:
One last nitpick: `caml_process_pending_callbacks` does not perform the urgent GC requests, but resets `caml_something_to_do`, meaning that if it is called directly, the urgent GC requests will be delayed until the next minor GC.

Could you either call `caml_check_urgent_gc` here instead of calling it in `caml_check_urgent_gc_and_callbacks`, or reset `caml_something_to_do` in `caml_check_urgent_gc_and_callbacks` instead of here?

Reply:
So I address this in the fixup commit I just pushed (I will squash it later). We cannot really do as you say, because the sole purpose of `caml_process_pending_callbacks` is to be called by people who make sure `caml_gc_dispatch` has been called just before or does not need to be called (`caml_alloc_small_dispatch` and others). So, I documented this usage, and checked it in the code: the use in `enter/leave_blocking_section` was indeed incorrect. I replace it with a direct call to signal handlers because:
* This is how it was done before
* It is not rare that signal handlers are non-allocating, so in case a minor collection or a major slice is pending, calling caml_check_urgent_gc_and_callbacks behaves differently by delaying the signal handling until gc operations are complete, for the sake of factorizing code. This is debatable, but not justified by the object of the various PRs in this thread.

Reply:
Thanks again for the nitpick by the way.

Reply:
What if `caml_process_pending_callbacks` called `caml_check_urgent_gc` too? Sure, this can end up in it being called twice, but do we care?

The way I was seeing this is that `caml_process_pending_callbacks` and `caml_process_pending_actions` were doing the same thing, except that the latter checked for `caml_something_to_do` (and hence was faster).